### ISSUE TITLE
About expiration performance

### ISSUE THREAD

**Issue by XieGuochao** [maypok86/otter#90]
I am the author of [Bucket-Based Expiration Algorithm: Improving Eviction Efficiency for In-Memory Key-Value Database](https://dl.acm.org/doi/fullHtml/10.1145/3422575.3422797) and today I find this project :) Really nice implementation in Go!

For curiosity, I wonder how you found this paper. Do you have a benchmark for the expiration? And is there any space I could contribute to this project?  (I have >5 years experiences with Go.)

**Comment by ben-manes** [comment]
I believe the implementation is a [port](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/TimerWheel.java) from Caffeine (Java, 2017). Were you aware of timing wheels? They’re very elegant, the [paper](https://dl.acm.org/doi/10.1145/41457.37504) discusses a few variations, and it’s neat that you may have rediscovered them. I wrote about this [here](https://highscalability.com/design-of-a-modern-cachepart-deux/) and included it in a Usenix Fast’20 talk ([slides](https://docs.google.com/presentation/u/0/d/1NlDxyXsUG1qlVHMl4vsUUBQfAJ2c2NsFPNPr2qymIBs/mobilepresent), [full slides](https://drive.google.com/file/d/1oHV8SjrdhSdwgZ6jegJh8Ysrbnzxvlfk/view?usp=drivesdk)). That may have spread the idea for use in a cache where I’ve seen it ported to go ([alt](https://github.com/Yiling-J/theine-go/blob/main/internal/timerwheel.go)), [rust](https://github.com/moka-rs/moka/blob/main/src/common/timer_wheel.rs) ([alt](https://github.com/Yiling-J/theine-core/blob/main/src/timerwheel.rs)), [typescript](https://github.com/aholstenson/timer-wheel), [c#](https://github.com/bitfaster/BitFaster.Caching/pull/516). I don’t know of any usages for a cache predating those, but I wouldn’t be surprised if it was as popular in low-level circles.

I have a [benchmark](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/jmh/java/com/github/benmanes/caffeine/cache/TimerWheelBenchmark.java) for an implementation sanity check and not as a comparison.

**Comment by maypok86** [comment]
Wow, this is unexpected.

> For curiosity, I wonder how you found this paper.

I wanted to use the algorithm from ristretto at first but there were several problems with it.

1. Several uncorrected bugs that lead to memory leaks under high contention loads.
2. I wanted to delete expired entries faster than ristretto does.

I decided to see how redis does it and I was interested in its old algorithm. Unfortunately, it is not proactive at all and I had the idea to combine ideas from the ristretto (buckets) and redis algorithms. But before I started doing anything, I googled the keywords and your paper ended up on the first lines.

> Do you have a benchmark for the expiration?

I checked the speed of some cases but I didn't seem to upload it to the public.

I have changed the implementation to timer wheel so far simply because it is more proactive and its disadvantages should not be noticeable on onheap caches. But I have plans to add the ability to pass a custom allocator to allow storing data outside the heap and I'm not sure that timer wheel will feel good on such caches. So maybe I'll have to get the old commits :).

P.S. Your paper is great!
